easybuild/easyblocks/generic/cmakemake.py

```diff
@@ -187,32 +198,49 @@ def configure_step(self, srcdir=None, builddir=None):
                 print_warning('Ignoring BUILD_SHARED_LIBS is set in configopts because build_shared_libs is set')
             self.cfg.update('configopts', '-DBUILD_SHARED_LIBS=%s' % ('ON' if build_shared_libs else 'OFF'))
 
-        env_to_options = {
-            'CC': 'CMAKE_C_COMPILER',
-            'CFLAGS': 'CMAKE_C_FLAGS',
-            'CXX': 'CMAKE_CXX_COMPILER',
-            'CXXFLAGS': 'CMAKE_CXX_FLAGS',
-            'F90': 'CMAKE_Fortran_COMPILER',
-            'FFLAGS': 'CMAKE_Fortran_FLAGS',
-        }
+        # If the cache does not exist CMake reads the environment variables
+        cache_exists = os.path.exists('CMakeCache.txt')
+        env_to_options = dict()
+
+        # Setting compilers is not required unless we want absolute paths
+        if self.cfg.get('abs_path_compilers', False) or cache_exists:
+            env_to_options.update({
+                'CC': 'CMAKE_C_COMPILER',
+                'CXX': 'CMAKE_CXX_COMPILER',
+                'F90': 'CMAKE_Fortran_COMPILER',
+            })
+        else:
+            # Set the variable which CMake uses to init the compiler using F90 for backward compatibility
+            fc = os.getenv('F90')
+            if fc:
+                setvar('FC', fc)
+
+        # Flags are read from environment variables already since at least CMake 2.8.0
+        if LooseVersion(get_software_version('CMake')) < LooseVersion('2.8.0') or cache_exists:
```

Comment:
We're making the assumption here that CMake is indeed listed as a (build) dependency, which was not the case before.

Prior to these changes, the configure step was failing with "`cmake: command not found`" (unless `cmake` was also available in the OS), but now we get a nasty crash:

```
ERROR: Traceback (most recent call last):
  File "/data/gent/400/vsc40023/easybuild_easy_installed/lib/python3.6/site-packages/easybuild_framework-4.6.1.dev0-py3.6.egg/easybuild/main.py", line 128, in build_and_install_software
    (ec_res['success'], app_log, err) = build_and_install_one(ec, init_env)
  File "/data/gent/400/vsc40023/easybuild_easy_installed/lib/python3.6/site-packages/easybuild_framework-4.6.1.dev0-py3.6.egg/easybuild/framework/easyblock.py", line 4077, in build_and_install_one
    result = app.run_all_steps(run_test_cases=run_test_cases)
  File "/data/gent/400/vsc40023/easybuild_easy_installed/lib/python3.6/site-packages/easybuild_framework-4.6.1.dev0-py3.6.egg/easybuild/framework/easyblock.py", line 3960, in run_all_steps
    self.run_step(step_name, step_methods)
  File "/data/gent/400/vsc40023/easybuild_easy_installed/lib/python3.6/site-packages/easybuild_framework-4.6.1.dev0-py3.6.egg/easybuild/framework/easyblock.py", line 3795, in run_step
    step_method(self)()
  File "/data/gent/400/vsc40023/easybuild_easy_installed/lib/python3.6/site-packages/easybuild_easyblocks-4.6.1.dev0-py3.6.egg/easybuild/easyblocks/generic/cmakemake.py", line 221, in configure_step
    if LooseVersion(get_software_version('CMake')) < LooseVersion('2.8.0') or cache_exists:
  File "/usr/lib64/python3.6/distutils/version.py", line 52, in __lt__
    c = self._cmp(other)
  File "/usr/lib64/python3.6/distutils/version.py", line 335, in _cmp
    if self.version == other.version:
AttributeError: 'LooseVersion' object has no attribute 'version'
```

We're also not taking into account that CMake may be filtered out as a dependency via `--filter-deps` (some people may prefer using the CMake that is provided via the OS), so we should have a fallback to determining the CMake version via `cmake --version` if `get_software_version` returns `None`.

@Flamefire Are you up for looking into this? If not, we should open an issue to avoid forgetting about this (we should handle this prior to the next EasyBuild release that includes the changes in this PR)

Reply:
I looked into this myself, see #2772